File: atoci/Arrays & Hashing/cpp/twoSum.cpp
#include<iostream>
#include<unordered_map>
#include<vector>

using namespace std; 

vector<int> twoSum(const vector<int>& nums, const int& target) {
    unordered_map<int, int> mymap;
    int lookup; 

    for (int i = 0; i < nums.size(); i++) {
        lookup = target - nums[i];

        if (mymap.find(lookup) != mymap.end()) {
            return {mymap[lookup], i};
        }
        mymap[nums[i]] = i;
    }
    
    return {};
}

ostream& operator<<(ostream& os, const vector<int>& nums) {
    for (const auto& val : nums) {
        os << val << " ";
    }
    return os; 
}


int main() {
    vector<int> result; 
    
    cout << twoSum({2,7,11,15}, 9) << endl;
    cout << twoSum({3,2,4}, 6) << endl;
    cout << twoSum({3,3}, 6) << endl; 
    return 0;
}

File: atoci/Arrays & Hashing/typescript/twoSum.ts
function twoSum(nums: number[], target:number): number[] {
    const mymap : {[key: number]: number} = {};
    var lookup : number = 0;

    for (let index = 0; index < nums.length; index++) {
        lookup = target - nums[index]

        if (lookup in mymap) {
            if (mymap[lookup] != index) {
                return [mymap[lookup], index]
            }
        }

        mymap[nums[index]] = index
    }

    return []
}

console.log(twoSum([2,7,11,15], 9))
console.log(twoSum([3,2,4], 6))
console.log(twoSum([3,3], 6))


